woocommerce/google-listings-and-ads
# Prefill target audience countries with suggestions

### Changes proposed in this Pull Request:

Part of issue #1116, based on PR #1131 using API developed by @nima-karimi.

In this PR, when users first arrive at Setup MC Step 2 Choose your audience page, we will pre-fill the target audience countries with suggestions retrieved from `GET /mc/target_audience/suggestions`.

The logic goes like this: 

1. First, we check if there is existing countries value in `GET /mc/target_audience`. If there is, this means users have already selected their own countries before, and we should not continue with pre-fill. If the countries value is null, this means users is visiting this page for the first time, and we should proceed with pre-fill.
2. Call `GET /mc/target_audience/suggestions` and fill it into the form.

The countries retrieved from suggestions are based on a combination of factors:

- Countries defined in WooCommerce shipping zones settings. If you select a region, all countries within the region would be included. If you select a state, the country of the state would be included.
- WooCommerce store country settings.
- All the included countries above would then be filtered against list of countries supported by Google Merchant Center. If all the countries above are not supported, you will have 0 countries (an empty array in the `countries` field in the API response).

See the demo video below for more detailed walkthrough.

**Note:** in this PR, we are only changing one file `useAutoSetLocationCountriesEffect.js`. In a subsequent PR #1146 (to fix UI loading flicker) that is based on this PR, there is a change of loading approach and the file is removed. Nevertheless, this PR is still worth reviewing and merging, in case #1146 is deemed undesirable.

### Screenshots:

📹 Demo video:

https://user-images.githubusercontent.com/417342/145185996-a0fdcd7d-290c-4a1c-9240-ffb81ee8371c.mov

### Detailed test instructions:

1. Setup your WooCommerce store country and shipping zones.
2. Go through account connections in Setup MC Step 1. Before clicking on the Continue button, open your browser devtools network panel to observe network requests.
3. Click on the Continue button to go to Step 2. You should see `GET /mc/target_audience` API call, which returns `countries: null`. Then, then should be `GET /mc/target_audience/suggestions` API call, which returns `countries` with array value containing country codes.
4. The country dropdown select UI should be pre-filled with the suggested countries.
5. There should be a `POST /mc/target_audience` to save the pre-filled country values. This is existing production behavior.


### Changelog entry

> Add - Pre-fill target audience countries with suggestions based on WooCommerce settings.
